amanuensis/resources/admin/admin_associated_user.py

```diff
@@ -25,10 +29,73 @@ def add_associated_users(users):
     with flask.current_app.db.session as session:
         associated_user_schema = AssociatedUserSchema(many=True)
         ret = []
-        for user in users:
-            ret.append(udm.add_associated_user(session, user["project_id"], user["email"] if "email" in user else None, user["id"] if "id" in user else None))
-        associated_user_schema.dump(ret)
-        return ret
+        users_with_email = [user for user in users if "email" in user]
+        users_with_id = [
+            user
+            for user in users
+            if "id" in user
+        ]
+        if not (users_with_email or users_with_id):
+            raise UserError("The associated user must have at least an email or id. Neither was found.")
+        user_emails = [user["email"] for user in users_with_email]
+        user_ids = [user["id"] for user in users_with_id]
+        associated_users = udm.get_associated_users(session, user_emails)
+        associated_users += udm.get_associated_users_by_id(session, user_ids)
+
+        associated_user_emails = [user.email for user in associated_users]
+        associated_user_ids = [user.user_id for user in associated_users]
 
 
+        user_emails = [user["email"] for user in users_with_email]
+        user_ids = [user["id"] for user in users_with_id]
 
+        fence_users = []
+        if user_emails:
+            fence_users.extend(fence.fence_get_users(config, usernames=user_emails)["users"])
+        if user_ids:
+            fence_users.extend(fence.fence_get_users(config, ids=user_ids)["users"])
+        if fence_users:
+            fence_users = [
+                {"email": user["name"], "id": user["id"]} for user in fence_users
+            ]
+
+        for fence_user in fence_users:
+            for user in users_with_email:
+                if user["email"] == fence_user["email"]:
+                    user["id"] = fence_user["id"]
+                    break
+
+        for fence_user in fence_users:
+            for user in users_with_id:
+                if user["id"] == fence_user["id"]:
+                    user["email"] = fence_user["email"]
+                    break
+        
+        users.clear()
+        users = users_with_email + users_with_id
```

Comment:
This also could create duplicates with the logic above.

Reply:
It checks for duplicates later on. But this is a bit confusing. I'm going to go ahead and see if I can refactor as you suggested earlier.

Reply:
I changed the code to either get users by id or email.